File: bluetooth/wifi/wifi-ssid-characteristic.js
var util = require('util');
var bleno = require('bleno');

var CHARACTERISTIC_NAME = 'SSID';

var WifiSsidCharacteristic = function(wifiState) {
    WifiSsidCharacteristic.super_.call(this, {
        uuid: '18c7042c-12da-49e8-845e-6086d18a81fa',
        properties: ['read', 'write'],
        secure: [],
        descriptors: [
            new bleno.Descriptor({
                uuid: '2901',
                value: CHARACTERISTIC_NAME,
            }),
            new bleno.Descriptor({
                uuid: '2904',
                value: new Buffer([0x04, 0x00, 0x27, 0x00, 0x01, 0x00, 0x00])
            }),
        ],
    });

    this._update = null;

    this.wifiState = wifiState;

}

util.inherits(WifiSsidCharacteristic, bleno.Characteristic);

WifiSsidCharacteristic.prototype.onReadRequest = function(offset, callback) {
    console.log('onReadRequest');
    if (offset) {
        console.log('onReadRequest offset');
        callback(this.RESULT_ATTR_NOT_LONG, null);
    } else {
        let responseData = new Buffer(this.wifiState.ssid);
        console.log("onReadRequest returning ", responseData);
        callback(this.RESULT_SUCCESS, responseData);
    }
}

WifiSsidCharacteristic.prototype.onWriteRequest = function(data, offset, withoutRespose, callback) {
    if (offset) {
        callback(this.RESULT_ATTR_NOT_LONG);
    } else if (data.length <= 0) {
        callback(this.RESULT_INVALID_ATTRIBUTE_LENGTH);
    } else {
        let str = data.toString();
        this.wifiState.set_ssid(str);
        console.log(data.toString());
        callback(this.RESULT_SUCCESS);
    }
}

module.exports = WifiSsidCharacteristic;

File: bluetooth/services/wifi_service/wifi_ssid_characteristic.py
from pybleno import *
import array

CHARACTERISTIC_NAME = "SSID"


class SSIDCharacteristic(Characteristic):
    def __init__(self, wifi_state):
        Characteristic.__init__(self, {
            'uuid': '3002A7D3-D8A4-4FEA-8174-1736E808C066',
            'properties': ['read', 'write'],
            'value': None,
            'descriptors': [
                Descriptor({
                    'uuid': '2901',
                    'value': bytes(CHARACTERISTIC_NAME, 'utf-8')
                }),
                Descriptor({
                    'uuid': '2904',
                    'value': array.array('B', [0x19, 0x00, 0x27, 0x00, 0x01, 0x00, 0x00])
                })
            ]
        })

        self._wifi_state = wifi_state

    def onReadRequest(self, offset, callback):
        if offset:
            callback(Characteristic.RESULT_ATTR_NOT_LONG, None)
        else:
            data = bytes(self._wifi_state.ssid, 'utf-8')
            callback(Characteristic.RESULT_SUCCESS, data)

    def onWriteRequest(self, data, offset, withoutResponse, callback):
        if offset:
            callback(Characteristic.RESULT_ATTR_NOT_LONG)
        elif len(data) <= 0:
            callback(Characteristic.RESULT_INVALID_ATTRIBUTE_LENGTH)
        else:
            ssid = data.decode("utf-8")
            self._wifi_state.set_ssid(ssid)
            callback(Characteristic.RESULT_SUCCESS)

